src/MonitorProfiler/Communication/CommandServer.cpp

```diff
@@ -0,0 +1,104 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "CommandServer.h"
+#include <thread>
+#include "../Logging/Logger.h"
+
+CommandServer::CommandServer(std::shared_ptr<ILogger> logger) : _shutdown(false), _logger(logger)
+{
+}
+
+HRESULT CommandServer::Start(const std::string& path, std::function<HRESULT(const IpcMessage& message)> callback)
+{
+    if (_shutdown.load())
+    {
+        return E_UNEXPECTED;
+    }
+
+    HRESULT hr;
+#if TARGET_WINDOWS
+    WSADATA wsaData;
+    int result = WSAStartup(MAKEWORD(2, 2), &wsaData);
+    if (result != 0)
+    {
+        return HRESULT_FROM_WIN32(result);
+    }
+#endif
+
+    _callback = callback;
+
+    IfFailLogRet_(_logger, _server.Bind(path));
+    _listeningThread = std::thread(&CommandServer::ListeningThread, this);
+    _clientThread = std::thread(&CommandServer::ClientProcessingThread, this);
+    return S_OK;
+}
+
+void CommandServer::Shutdown()
+{
+    bool shutdown = false;
+    if (_shutdown.compare_exchange_strong(shutdown, true))
+    {
+        _clientQueue.Complete();
+        _server.Shutdown();
+
+        _listeningThread.join();
+        _clientThread.join();
+    }
+}
+
+void CommandServer::ListeningThread()
+{
+    while (true)
+    {
+        std::shared_ptr<IpcCommClient> client;
+        HRESULT hr = _server.Accept(client);
+        if (FAILED(hr))
+        {
+            break;
+        }
+
+        IpcMessage message;
+
+        //Note this can timeout if the client doesn't send anything
+        hr = client->Receive(message);
+        if (FAILED(hr))
+        {
+            _logger->Log(LogLevel::Error, _T("Unexpected error when receiving data: 0x%08x"), hr);
+            continue;
+        }
+
+        IpcMessage response;
+        response.MessageType = SUCCEEDED(hr) ? MessageType::OK : MessageType::Error;
+        response.Parameters = hr;
+
+        hr = client->Send(response);
+        if (FAILED(hr))
+        {
+            _logger->Log(LogLevel::Error, _T("Unexpected error when sending data: 0x%08x"), hr);
+            continue;
+        }
+
```

Comment:
Should there be a call to `client.Shutdown()` here?

Reply:
That would probably be cleaner, yes. My original approach involved multiple clients and a fully blocking socket so shutdown was a much stricter requirement.